projects/plugins/jetpack/_inc/client/components/module-settings/with-module-settings-form-helpers.jsx

```diff
@@ -17,12 +17,17 @@ export function withModuleSettingsFormHelpers( InnerComponent ) {
 		};
 
 		onOptionChange = event => {
-			const optionName = event.target.name;
+			let optionName = event.target.name;
 			let optionValue;
 			// Get the option value from the `checked` property if present.
 			if ( event.target.type === 'checkbox' ) {
 				optionValue =
 					typeof event.target.checked !== 'undefined' ? event.target.checked : event.target.value;
+			} else if ( optionName.includes( '.' ) ) {
+				// We define the use of an object in a form field name like "object.member", ie: "subscription_options.welcome".
+				// Then we create an object with the setting name and the subsetting name. This is useful for nested settings.
+				const [ settingName, subSettingName ] = optionName.split( '.' );
+				optionName = { [ settingName ]: { [ subSettingName ]: event.target.value } };
```

Comment:
The above lines together with `onChange={ props.onOptionChange }` in `subscription-settings.jsx`.

Instead of calling `props.onOptionChange` directly, perhaps you could hook it to an intermediate function that would construct a new `subscription_options` object with the new welcome message, set that object back to the `event.target.value` property, and then pass it back to `onOptionChange`.

With that, you could skip writing the additional handling of nested objects on this `onOptionChange` part.

Reply:
I agree with Yan.

Reply:
I see your points. I agree there is a lot of standardization between modules we can do. But, with the settings module, it seems like originally, it was thought to manage objects, but one of the functions didn't count with them. Let me explain:
`onOptionChange` prepares the pair `optionName` / `optionValue` and passes it to `updateFormStateOptionValue`. Before my changes, only checkboxes or strings were expected in `onOptionChange,` but if you take a look at the code in `updateFormStateOptionValue,` it was expecting the `optionName` to be a string or... object. In fact, if that variable is a string, it converts it to an object to be merged with the options set in the state.
I mean that, clearly, `updateFormStateOptionValue` was prepared to handle objects already, but `onOptionChange` was not. My change made `onOptionChange` ready for objects, too.

Reply:
> Before my changes, only checkboxes or strings were expected in onOptionChange [...] I mean that, clearly, updateFormStateOptionValue was prepared to handle objects already, but onOptionChange was not. My change made onOptionChange ready for objects, too.

The `onOptionChange` function is implicitly ready for objects as long as the `event.target.value` contains an object.

Whereas the current implementation explicitly handles objects by determining whether there is a `.` in the option name is leaving room for further ambiguity, e.g.:
- Is the dot meant for nested object or is the dot part of the property name?
- What if I want to store multiple level of nested objects, would the multiple dot notation work?

> but if you take a look at the code in updateFormStateOptionValue, it was expecting the optionName to be a string or... object. In fact, if that variable is a string, it converts it to an object to be merged with the options set in the state.

Linking to this [function here](https://github.com/Automattic/jetpack/blob/e272c9570a01c1c79eecc78925fa84a785241253/projects/plugins/jetpack/_inc/client/components/module-settings/with-module-settings-form-helpers.jsx#L46C3-L46C29). While the param is called `optionMaybeOptions`, which excepts either string or an object, for our use case, when calling in from `onOptionChange`, passing a string `subscription_options` with an object for the 2nd param would work perfectly just fine.

By skipping the explicitly object handling on `onOptionChange`, it would be reducing redundant work and future maintenance.

Reply:
You are right. I'll do the changes today 👍